src/main/java/net/neoforged/neoforge/capabilities/BlockCapabilityCache.java

```diff
@@ -0,0 +1,120 @@
+     *
+     * @param capability           the capability
+     * @param level                the level
+     * @param pos                  the position
+     * @param context              extra context for the query
+     * @param isValid              a function to check if the listener still wants to receive notifications.
+     *                             A typical example would be {@code () -> !this.isRemoved()} for a block entity
+     *                             that should not receive invalidation notifications anymore once it is removed.
+     * @param invalidationListener the invalidation listener. Will be called whenever the capability of the cache might have changed.
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context, BooleanSupplier isValid, Runnable invalidationListener) {
+        Objects.requireNonNull(capability);
+        Objects.requireNonNull(isValid);
+        Objects.requireNonNull(invalidationListener);
+        pos = pos.immutable();
+
+        var cache = new BlockCapabilityCache<>(capability, level, pos, context, isValid, invalidationListener);
+        level.registerCapabilityListener(pos, cache.listener);
+        return cache;
+    }
+
+    private final BlockCapability<T, C> capability;
+    private final ServerLevel level;
+    private final BlockPos pos;
+    private final C context;
+
+    private boolean cacheValid = false;
+    @Nullable
+    private T cachedCap = null;
+
+    private boolean canQuery = true;
+    private final ICapabilityInvalidationListener listener;
+
+    private BlockCapabilityCache(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context, BooleanSupplier isValid, Runnable invalidationListener) {
+        this.capability = capability;
+        this.level = level;
+        this.pos = pos;
+        this.context = context;
+
+        this.listener = () -> {
+            if (!cacheValid)
+                return true; // cache was already invalidated...
+
+            canQuery = false;
+            cacheValid = false;
+
+            if (isValid.getAsBoolean()) {
+                invalidationListener.run();
+                canQuery = true;
+                return true;
+            } else {
+                return false;
+            }
+        };
+    }
+
+    public ServerLevel level() {
+        return level;
+    }
+
+    public BlockPos pos() {
+        return pos;
+    }
+
+    public C context() {
+        return context;
+    }
+
+    /**
+     * Gets the capability instance, or null if the capability is not present.
+     */
+    @Nullable
+    public T getCapability() {
+        if (!canQuery)
+            throw new IllegalStateException("Do not call getCapability on an invalid cache or from the invalidation listener!");
+
+        if (!cacheValid) {
+            cachedCap = level.getCapability(capability, pos, context);
```

Comment:
As discussed [on discord](https://discord.com/channels/313125603924639766/1135328752658829312/1177665948291969094) we should check if the position is loaded and if it isn't set `cachedCap` to `null` instead of querying the capability from the level.

Reply:
Added an is loaded check.